framework/include/constraints/AutomaticMortarGeneration.h

```diff
@@ -0,0 +1,252 @@
+  // nodes are "matching".
+  // In this configuration:
+  //
+  //    A     B
+  // o-----o-----o  (slave orientation ->)
+  //       |
+  //       v
+  // ------x------ (master orientation <-)
+  //    C     D
+  //
+  // The entries in the map should be:
+  // (Elem A, Node 1) -> (Elem C, xi^(2)=-1)
+  // (Elem B, Node 0) -> (Elem D, xi^(2)=+1)
+  std::map<std::pair<const Node *, const Elem *>, std::pair<Real, const Elem *>>
+      slave_node_and_elem_to_xi2_master_elem;
+
+  // Same type of container, but for mapping
+  // (Master Node, Master Elem) -> (xi^(1), Slave Elem)
+  // where they are inverse-projected along the nodal normal
+  // direction.
+  std::map<std::pair<const Node *, const Elem *>, std::pair<Real, const Elem *>>
+      master_node_and_elem_to_xi1_slave_elem;
+
+  // 1D Mesh of mortar segment elements which gets built by the call
+  // to build_mortar_segment_mesh().
+  Mesh mortar_segment_mesh;
+
+  // Map between Elems in the mortar segment mesh and their info
+  // structs. This gets filled in by the call to
+  // build_mortar_segment_mesh().
+  std::map<const Elem *, MortarSegmentInfo> msm_elem_to_info;
+
+  // Keeps track of the mapping between lower-dimensional elements and
+  // the side_id of the interior_parent which they are.
+  std::map<const Elem *, unsigned int> lower_elem_to_side_id;
+
+  // The amount by which elements in the boundary subdomains
+  // are offset from their respective interior parent's subdomain ids.
+  const subdomain_id_type boundary_subdomain_id_offset = 1000;
+
+  // A list of master/slave subdomain id pairs corresponding to each
+  // side of the mortar interface.
+  std::vector<std::pair<subdomain_id_type, subdomain_id_type>> master_slave_subdomain_id_pairs;
+
+  // The slave/master lower-dimensional boundary subdomain ids are the
+  // slave/master *boundary* ids offset by the value above.
+  std::set<subdomain_id_type> slave_boundary_subdomain_ids;
+  std::set<subdomain_id_type> master_boundary_subdomain_ids;
+
+  // Size of the largest slave side lower-dimensional element added to
+  // the mesh. Used in plotting convergence data.
+  Real h_max;
+
+  // Used by the AugmentSparsityOnInterface functor to determine
+  // whether a given Elem is coupled to any others across the gap, and
+  // to explicitly set up the dependence between interior_parent()
+  // elements on the slave side and their lower-dimensional sides
+  // which are on the interface. This latter type of coupling must be
+  // explicitly declared when there is no master_elem for a given
+  // mortar segment and you are using e.g.  a P^1-P^0 discretization
+  // which does not induce the coupling automatically.
+  std::unordered_multimap<dof_id_type, dof_id_type> mortar_interface_coupling;
+
+  // Container for storing the nodal normal vector associated with each slave node.
+  std::unordered_map<const Node *, Point> slave_node_to_nodal_normal;
```

Comment:
OK - so you used `unordered_map` here. I don't immediately see why we can't in the other places I've indicated, am I missing something?

Reply:
K, switched all maps that could be switched to `unordered`